network/kademlialoadbalancer.go

```diff
@@ -0,0 +1,360 @@
+	for i, res := range resources {
+		peer := res.(*Peer)
+		peers[i].Peer = peer
+		peers[i].Use = func() {
+			klb.resourceUseStats.addUse(peer)
+		}
+	}
+	return peers
+}
+
+func (klb *KademliaLoadBalancer) getPeersForPo(base []byte, po int) []LBPeer {
+	resources := make([]Resource, 0)
+	klb.kademlia.EachBinDesc(base, po, func(bin *PeerBin) bool {
+		if bin.ProximityOrder == po {
+			return bin.PeerIterator(func(entry *entry) bool {
+				resources = append(resources, entry.conn)
+				return true
+			})
+		} else {
+			return true
+		}
+	})
+	return klb.resourcesToLbPeers(resources)
+}
+
+// Resource Use Stats
+
+// resourceUseStats can be used to count uses of resources. A Resource is anything with a Key()
+type resourceUseStats struct {
+	resourceUses map[string]int
+	waiting      map[string]chan struct{}
+	lock         sync.RWMutex
+}
+
+type Resource interface {
+	Key() string
+}
+
+// Adding Resource interface to Peer
+func (d *Peer) Key() string {
+	return hexutil.Encode(d.Address())
+}
+
+type ResourceCount struct {
+	resource Resource
+	count    int
+}
+
+func newResourceLoadBalancer() *resourceUseStats {
+	return &resourceUseStats{
+		resourceUses: make(map[string]int),
+		waiting:      make(map[string]chan struct{}),
+	}
+}
+
+func (lb *resourceUseStats) sortResources(resources []Resource) []Resource {
+	sorted := make([]Resource, len(resources))
+	resourceCounts := lb.getAllUses(resources)
+	sort.Slice(resourceCounts, func(i, j int) bool {
+		return resourceCounts[i].count < resourceCounts[j].count
+	})
+	for i, resourceCount := range resourceCounts {
+		sorted[i] = resourceCount.resource
+	}
+	return sorted
+}
+
+func (lbp ResourceCount) String() string {
+	return lbp.resource.Key() + ":" + strconv.Itoa(lbp.count)
+}
+
+func (lb *resourceUseStats) dumpAllUses() map[string]int {
+	lb.lock.RLock()
+	defer lb.lock.RUnlock()
+	dump := make(map[string]int)
+	for k, v := range lb.resourceUses {
+		dump[k] = v
+	}
+	return dump
+}
+
+func (lb *resourceUseStats) getAllUses(resources []Resource) []ResourceCount {
+	peerUses := make([]ResourceCount, len(resources))
+	for i, resource := range resources {
+		peerUses[i] = ResourceCount{
+			resource: resource,
+			count:    lb.getUses(resource),
+		}
+	}
+	return peerUses
+}
+
+func (lb *resourceUseStats) getUses(keyed Resource) int {
+	return lb.getKeyUses(keyed.Key())
+}
+
+func (lb *resourceUseStats) getKeyUses(key string) int {
+	lb.lock.RLock()
+	defer lb.lock.RUnlock()
+	return lb.resourceUses[key]
+}
+
+func (lb *resourceUseStats) addUse(resource Resource) int {
+	lb.lock.Lock()
+	defer lb.lock.Unlock()
+	log.Debug("Adding use", "key", resource.Key()[:4])
+	key := resource.Key()
+	lb.resourceUses[key] = lb.resourceUses[key] + 1
+	return lb.resourceUses[key]
+}
+
+// Used for testing. As peer resource initialization is asynchronous we need a way
+// to know that the initial uses has been initialized for a new peer
+func (lb *resourceUseStats) waitKey(key string) {
```

Comment:
Well, I can't remove this code from here, as the mechanism to subscribe to a peer must be inside the initKey method (that is not test-only code).
Even though it will be used only in testing for the moment, maybe in the future it could be used for checking that some peer has been initialized in the load balancer.

Reply:
I have removed teh used for testing comment and keep the method.